File: 0094/iterative.cpp
#include <cassert>
#include <iostream>
#include <stack>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  vector<int> inorderTraversal(TreeNode* root) {
    vector<int> ret;
    stack<TreeNode*> node_stack;
    while (root || !node_stack.empty()) {
      while (root) {
        node_stack.push(root);
        root = root->left;
      }
      root = node_stack.top();
      node_stack.pop();
      ret.push_back(root->val);
      root = root->right;
    }
    return ret;
  }
};

int main() {
  Solution sol;
  TreeNode* root;
  vector<int> got, expected;

  root = CreateTreeNode("1 # 2 # # 3 ;");
  got = sol.inorderTraversal(root);
  expected = {1,3,2};
  assert(got == expected);
  DeleteTreeNode(root);

  //     1
  //   2   3
  //  4 5 6 7
  root = CreateTreeNode("1 2 3 4 5 6 7 ;");
  got = sol.inorderTraversal(root);
  expected = {4,2,5,1,6,3,7};
  assert(got == expected);
  DeleteTreeNode(root);

  got = sol.inorderTraversal(nullptr);
  assert(got.size() == 0);
}

File: 0100/iterative.py
from typing import *


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        que = []
        que.append((p, q))
        while len(que) > 0:
            p, q = que.pop(0)
            if p is None and q is None:
                continue
            elif (p is None and q is not None) or (p is not None and q is None):
                return False
            if p.val != q.val:
                return False
            que.append((p.left, q.left))
            que.append((p.right, q.right))
        return True


if __name__ == '__main__':
    sol = Solution()

    p = TreeNode(1)
    p.left = TreeNode(2)
    p.right = TreeNode(3)
    q = TreeNode(1)
    q.left = TreeNode(2)
    q.right = TreeNode(3)
    assert sol.isSameTree(p, q)

    p = TreeNode(1)
    p.left = TreeNode(2)
    q = TreeNode(1)
    q.right = TreeNode(2)
    assert not sol.isSameTree(p, q)

    p = TreeNode(1)
    p.left = TreeNode(2)
    p.right = TreeNode(1)
    q = TreeNode(1)
    q.left = TreeNode(2)
    q.right = TreeNode(1)
    assert sol.isSameTree(p, q)
